dkenv/dkenv.go

```diff
@@ -0,0 +1,206 @@
+package dkenv
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"math/rand"
+	"os"
+	"regexp"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+var (
+	apiVersions = map[string]string{
+		"1.12": "1.0.1",
+		"1.13": "1.1.2",
+		"1.14": "1.2.0",
+		"1.15": "1.3.3",
+		"1.16": "1.4.1",
+		"1.17": "1.5.0",
+		"1.18": "1.6.0",
+		"1.19": "1.7.1",
+		"1.20": "1.8.3",
+		"1.21": "1.9.1",
+	}
+)
+
+type Dkenv struct {
+	DkenvDir string
+	BinDir   string
+}
+
+func New(dkenvDir, binDir string) *Dkenv {
+	return &Dkenv{
+		DkenvDir: dkenvDir,
+		BinDir:   binDir,
+	}
+}
+
+func (d *Dkenv) ListAction() error {
+	installed, err := d.listInstalled()
+	if err != nil {
+		return err
+	}
+
+	if len(installed) == 0 {
+		log.Warning("No installed Docker binaries found!")
+		return nil
+	} else {
+		log.Infof("Found %v installed docker binaries", len(installed))
+		log.Info("") // blank line, for the pretty
+	}
+
+	for i := 0; i < len(installed); i++ {
+		log.Infof("%v: %v", i+1, installed[i])
+	}
+
+	return nil
+}
+
+func (d *Dkenv) FetchVersionAction(version string, api bool) error {
+	clientVersion := version
+
+	if api {
+		var err error
+
+		clientVersion, err = ApiToVersion(version)
+		if err != nil {
+			return err
+		}
+
+		log.Infof("Found client '%v' for API version '%v'", clientVersion, version)
+	}
+
+	if !d.isInstalled(clientVersion) {
+		// Attempt to download it
+		log.Infof("Docker version %v not found - attempting to download...", clientVersion)
+
+		if err := d.DownloadDocker(clientVersion); err != nil {
+			return err
+		}
+	} else {
+		log.Infof("Docker version %v already installed!", clientVersion)
+	}
+
+	// Update symlink
+	if err := d.UpdateSymlink(clientVersion); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Return a list of found docker binaries
+func (d *Dkenv) listInstalled() ([]string, error) {
+	fileList, err := ioutil.ReadDir(d.DkenvDir)
+	if err != nil {
+		return nil, err
+	}
+
+	found := make([]string, 0)
+
+	for _, filename := range fileList {
+		match, _ := regexp.MatchString(`docker-.+`, filename.Name())
+		if match {
+			found = append(found, filename.Name())
+		}
+	}
+
+	return found, nil
+}
+
+func (d *Dkenv) isInstalled(version string) bool {
+	if _, err := os.Stat(d.DkenvDir + "/docker-" + version); err == nil {
+		return true
+	}
+
+	return false
+}
+
+// Symlink docker to a dkenv docker binary
+//
+// - If destination exists AND is *not* a symlink:
+//     - backup destination file
+//     - create symlink
+// - If destination exists AND *is* a symlink:
+//     - check if the symlink already matches destination (noop if it does)
+//     - remove and overwrite if symlink is incorrect
+// - If destination does not exist:
+//     - create symlink
+//
+// TODO This will be a pain to test - refactor when possible.
+func (d *Dkenv) UpdateSymlink(version string) error {
```

Comment:
This is really meh.. would like to update this prior to merge; it'll also make testing it *much* simpler.